### PR TITLE
Bamboo-based spike traps

### PR DESCRIPTION
**Pull request by M-Earthfire**
[Hydroponics][Feature][Input-Wanted]

## About the PR 

This PR adds spike traps into the game.
Spike trap sprite 2 Spike trap packet sprite 2
Spike traps are created by using cable coils on bamboo stalks, consuming 3 bamboo stalks and 4 units of cable coil. After a 5 second action bar, a spike trap is created.

Arming spike traps is a quick action with a 2 second action bar by using the assembled trap in your hand. Once armed, any person who moves into the trap and is above walking speed suffers 9 damage and a 2 seconds stun while being forced onto the ground.

If a human slips or is thrown into a tile with the trap in place, their flight is stopped, the trap is destroyed and the person in question suffers 20 damage and a 3 second stun. All damage from the trap is reduced by "chest"-melee-resistance of the victim.

The trap can be laced with chemicals, both in armed and unarmed form. A trap can hold up to 15 chemicals and injects a person with chemicals up to half the damage dealt.

The trap can be disarmed and disassembled with wirecutters, with a 2 and 1 second action bar respectively.

## Why's this needed? 

Gives botany another angle to apply the chemicals they have access to. Also, this promotes "trapped jungles" in maintance for creative traitors.

In contrast to rafflesias or stinging nettles, their chemical application ignores chemprot and instead checks the melee resistance of the victim. This makes the trap a real threat to crewmember while the "on step"-effect of the trap is a nuisance to sec and not able to inject actually deadly amount of chemicals into them.

This makes the trap a supplement to other deadly botany items and not a replacement or contender.

## Changelog 

```changelog
(u)Lord_Earthfire
(*)Adds spike traps made out of bamboo and cables. Lace them with chemicals and trap the hallways like you always wanted!
```

### CODE REVIEW

File: code/obj/item/Bamboo_spike_trap.dm [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+				/obj/item/bamboo_spike_trap/proc/disarm,\list(user), W.icon, W.icon_state, "[user] finishes disarming [src]")
+				actions.start(action_bar, user)
+				return
+			else
+				if (ON_COOLDOWN(user, "dismantling_spike_trap", user.combat_click_delay))
+					return
+				playsound(src.loc, 'sound/items/Scissor.ogg', 60)
+				user.visible_message("[user] starts breaking down [src]...")
+				var/datum/action/bar/icon/callback/action_bar = new /datum/action/bar/icon/callback(user, src, src.break_down_time,
+				/obj/item/bamboo_spike_trap/proc/break_down,\list(user), W.icon, W.icon_state, "[user] finishes breaking down [src]")
```
Comment (TobleroneSwordfish) [inline code/obj/item/Bamboo_spike_trap.dm:54]: this should be indented if you want to break it over two lines (same with other cases)

File: code/obj/item/Bamboo_spike_trap.dm [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+				boutput(User, "<span class='alert'>A trap is already placed here!</span>")
+			return
+		if (!src.armed)
+			logTheThing(LOG_COMBAT, User, "armed a spike trap at [src.loc]")
+			set_icon_state("Bamboo_Spike_Trap-Assembled")
+			if (User)
+				User.drop_item(src)
+			src.armed = TRUE
+			src.anchored = TRUE
+		return
```
Comment (TobleroneSwordfish) [inline code/obj/item/Bamboo_spike_trap.dm:96]:
```suggestion
```

File: code/obj/item/Bamboo_spike_trap.dm [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+		var/trap_occupied = 0
+		for(var/obj/item/B in get_turf(src))
+			if (istype(B, /obj/item/bamboo_spike_trap))
+				var/obj/item/bamboo_spike_trap/BM = B
+				if (BM.armed)
+					trap_occupied = 1
+		if (trap_occupied)
+			if (User)
+				boutput(User, "<span class='alert'>A trap is already placed here!</span>")
+			return
```
Comment (TobleroneSwordfish) [inline code/obj/item/Bamboo_spike_trap.dm:88]: move the message and return into the loop so you stop when the trap is found and also don't need the `trap_occupied` var

File: code/obj/item/Bamboo_spike_trap.dm [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+		if (!src.armed)
+			logTheThing(LOG_COMBAT, User, "armed a spike trap at [src.loc]")
+			set_icon_state("Bamboo_Spike_Trap-Assembled")
+			if (User)
+				User.drop_item(src)
```
Comment (TobleroneSwordfish) [inline code/obj/item/Bamboo_spike_trap.dm:93]:
```suggestion
			User?.drop_item(src)
```
also the uppercase on `User` looks weird

File: code/obj/item/Bamboo_spike_trap.dm [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+	proc/disarm(mob/User)
+		if (!src)
+			return
+		if (src.armed)
+			set_icon_state("Bamboo_Spike_Trap-Packet")
+			logTheThing(LOG_COMBAT, User, "disarmed a spike trap at [src.loc]")
+			src.armed = FALSE
+			src.anchored = FALSE
+		return
```
Comment (TobleroneSwordfish) [inline code/obj/item/Bamboo_spike_trap.dm:106]:
```suggestion
```

File: code/obj/item/Bamboo_spike_trap.dm [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+	hitby(atom/movable/AM, datum/thrown_thing/thr)
+		..()
+		if (src.armed && ishuman(AM) && AM.throwing)
+			var/mob/living/carbon/human/victim = AM
+			//crashes into the trap when being thrown/slipped at it
+			victim.visible_message("<span class='alert'><B>[victim] crashes into the spike trap!</B></span>",\
+			"<span class='alert'><B>You crash into the spike trap!</B></span>")
+			crash_into(victim)
+			qdel(src) //if crashed into, destroys the trap
+			return
```
Comment (TobleroneSwordfish) [inline code/obj/item/Bamboo_spike_trap.dm:129]:
```suggestion
```

File: code/obj/item/Bamboo_spike_trap.dm [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+	attack_self(mob/User as mob)
+		if (!src.armed)
+			var/trap_occupied = 0
+			for(var/obj/item/B in get_turf(src))
+				if (istype(B, /obj/item/bamboo_spike_trap))
+					var/obj/item/bamboo_spike_trap/BM = B
+					if (BM.armed)
+						trap_occupied = 1
```
Comment (TobleroneSwordfish) [inline code/obj/item/Bamboo_spike_trap.dm:138]: same as above

File: code/obj/item/Bamboo_spike_trap.dm [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+		src.trap_damage(victim, src.crashed_force)
+		playsound(victim.loc, 'sound/impact_sounds/Generic_Hit_Heavy_1.ogg', 80, 1)
+		victim.UpdateDamageIcon()
+
+
+	proc/step_on(mob/living/carbon/human/victim as mob)
+		if (!src || !victim || !src.armed)
+			return
+		logTheThing(LOG_COMBAT, victim, "stepped into [src] at [log_loc(src)].")
+		victim.changeStatus("weakened", src.armed_weakened)
```
Comment (TobleroneSwordfish) [inline code/obj/item/Bamboo_spike_trap.dm:178]: stunned and weakened do exactly the same thing here with the same duration, curious why you're using two different effects?

File: code/obj/item/Bamboo_spike_trap.dm [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+	//This is copied out of runetrap.dm and modified to work with like the banana peel
+	//So, if you can slip over a banana peel (it has ignore_actual_delay checked), you should normally get impaled by the trap
+	proc/checkRun(var/mob/M)	//If we are above walking speed, this triggers
+		if(!M)
+			return
+		var/slip_delay = BASE_SPEED_SUSTAINED + WALK_DELAY_ADD
+		var/movement_delay_real = max(M.movement_delay(get_step(M,M.move_dir), 0),world.tick_lag)
+		if (movement_delay_real < slip_delay)
+			return TRUE
```
Comment (TobleroneSwordfish) [inline code/obj/item/Bamboo_spike_trap.dm:207]: this should probably just be a `mob` proc?

File: code/atom/throwing.dm [hunk 10]
```diff
@@ -25,6 +25,12 @@
 				if(!A.Cross(src))
 					src.throw_impact(A, thr)
 					. = TRUE
+			//Would be an idea to move all these checks into its own proc so non-humans don't need to check for this
+			if(ishuman(src) && istype(A, /obj/item/bamboo_spike_trap))
+				var/obj/item/bamboo_spike_trap/M = A
+				if(M.armed)
+					src.throw_impact(M, thr)
+					. = TRUE
```
Comment (TobleroneSwordfish) [inline code/atom/throwing.dm:33]: can't this just go in `bamboo_spike_trap/Crossed`? or is that not called for thrown things?

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]
These are really cool, though their interactions with ricin and potassium-ice mixtures might be a little strong given how mass producible they are.

File: code/atom/throwing.dm (hunk 10 above)
Comment (M-Earthfire) [reply code/atom/throwing.dm:33]: It is theoretically called for thrown things iirc. The thing is i cannot stop the throw within crossed iirc. I made it like so so the person cannot be thrown into 3-4 traps with a single throw. If there is a proc that i can call to stop the throw withing bamboo_spike_trap/crossed, i will move it there of course.

File: code/obj/item/Bamboo_spike_trap.dm (hunk 8 above)
Comment (M-Earthfire) [reply code/obj/item/Bamboo_spike_trap.dm:174]: If they do, i should probably make some changes to the duration. I thought with weakened here you are able to move while forced on the ground (honestly, completely forgot to test that)

File: code/obj/item/Bamboo_spike_trap.dm (hunk 9 above)
Comment (M-Earthfire) [reply code/obj/item/Bamboo_spike_trap.dm:203]: A check like that is in multiple places. Like runtrap.dm has one version that is copied from mob/proc/slip and this one uses a different version that is also copied from there. Turning it in it's own mob/proc is probably a good idea and makes it more flexible for other functions as well

Comment (M-Earthfire) [reply code/obj/item/Bamboo_spike_trap.dm:203]: While looking at that, it actually doesnt catch when someone runs... will change that real quick...

File: code/obj/item/Bamboo_spike_trap.dm (hunk 8 above)
Comment (M-Earthfire) [reply code/obj/item/Bamboo_spike_trap.dm:187]: Yeah, they are working completely the same. I make them both the weakened condition and reduce the duration when stepped into.

**Comment by M-Earthfire** [comment]
I can understand the balance concerns, and potassium+ice is the combo i am most concerned about. That's why i edited in the [Input-wanted] tag. But i edited after the checks ran, so sadly the bots didnt picked it up and made a forum threat.

If potassium+ice becomes a problem, i would probably add that the chem container of the  trap at least heats newly added chems to the temperature of the tile it is currently. So you need some work like preparing with cryostylane+oxygen beforehand.

File: code/obj/item/Bamboo_spike_trap.dm (hunk 9 above)
Comment (M-Earthfire) [reply code/obj/item/Bamboo_spike_trap.dm]: the whole thing was moved to be a mob proc.

**Comment by ZeWaka** [comment]
We talked about this and we don't really want this specific theme in the game. We'd be much more open to something similar but more botany-focused. For example, using the head from a maneater plant to make a man-eater beartrap. Or pitcher plants with chems. Tons of ideas, really.

**Comment by M-Earthfire** [comment]
I can certainly retool the code and make new sprites to shift the item to a different topic. Will do some brainstorming and then come up with another design for a mechanical simikar trap with a different theme

**Pull request closed without merging**